test-suite/testkit/src/api.rs

```diff
@@ -378,3 +386,75 @@ fn create_test_server(aggregator: ApiAggregator) -> TestServer {
     info!("Test server created on {}", server.addr());
     server
 }
+
+/// The custom implementation of the test server, because there is an error in the default
+/// implementation. It does not wait for the http server thread to complete during drop.
+struct TestServer {
+    addr: net::SocketAddr,
+    backend: Addr<Server>,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl TestServer {
+    /// Start new test server with application factory
+    fn with_factory<F, H>(factory: F) -> Self
+    where
+        F: Fn() -> H + Send + Clone + 'static,
+        H: IntoHttpHandler + 'static,
+    {
+        let (tx, rx) = mpsc::channel();
+
+        // run server in separate thread
+        let handle = thread::spawn(move || {
+            let sys = System::new("actix-test-server");
+            let tcp = net::TcpListener::bind("127.0.0.1:0").unwrap();
+            let local_addr = tcp.local_addr().unwrap();
+
+            let srv = HttpServer::new(factory)
+                .disable_signals()
+                .listen(tcp)
+                .keep_alive(5)
+                .workers(1)
+                .start();
+
+            tx.send((System::current(), local_addr, srv)).unwrap();
+            sys.run();
+        });
+
+        let (system, addr, backend) = rx.recv().unwrap();
+        System::set_current(system);
+
+        Self {
+            addr,
+            backend,
+            handle: Some(handle),
+        }
+    }
+
+    /// Construct test server url.
+    fn url(&self, uri: &str) -> String {
+        if uri.starts_with('/') {
+            format!("http://localhost:{}{}", self.addr.port(), uri)
+        } else {
+            format!("http://localhost:{}/{}", self.addr.port(), uri)
+        }
+    }
+
+    /// Construct test server url.
+    fn addr(&self) -> net::SocketAddr {
+        self.addr
+    }
+}
+
+impl Drop for TestServer {
```

Comment:
Nit: Is it possible to add a test that the socket address becomes free after `TestServer` is dropped?

Reply:
I think that it doesn't make much sense. This code is necessary only to avoid hidden issues under the `actix-web` hood.